### PR TITLE
feat: update the docs for edge function testing

### PR DESCRIPTION
**Pull request by mansueli**
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES

Based on the contents from: @kallebysantos 

https://github.com/kallebysantos/play.supabase.unit-tests

## What kind of change does this PR introduce?

Bug fix, feature, docs update, ...

## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

## Release Notes

* **Documentation**
  * Completely redesigned the unit and integration testing guide for Deno-based Supabase Edge Functions using a concrete example
  * Added a recommended test folder structure and example business-rule module
  * Introduced BDD-style unit testing patterns with runnable commands
  * Expanded end-to-end integration testing guidance, including mocking of outgoing REST calls
  * Updated test utilities for environment setup and JWT generation
  * Refreshed best practices and resources sections

### CODE REVIEW

**Review by daltino — COMMENTED** [review]
Great work on this docs update! The expanded content is a significant improvement over the previous version, adding much more practical guidance for developers testing Edge Functions. Here are my observations and suggestions:

---

## Overall Impressions

This is a substantial and valuable expansion of the unit testing documentation. Going from ~106 lines to ~403 additions shows a real commitment to helping developers understand testing strategies. The structure (unit tests → integration tests → mocking) is logical and easy to follow.

---

## Suggestions & Questions

### 1. The heading `# Testing your Edge Functions` may conflict with front matter

The file already has `title`, `description`, and `subtitle` in the front matter. Adding a top-level `#` heading inside an `.mdx` file that uses front matter for title rendering may cause a **duplicate heading** in the rendered output depending on how the docs site handles it. Most Supabase docs pages render the `title` from front matter automatically.

```suggestion
- # Testing your Edge Functions
-
- Testing is an essential step...
+ Testing is an essential step...
```

Check how other `.mdx` pages in `apps/docs/content/guides/` handle this — most likely the `#` heading should be removed.

---

### 2. Verify all code examples actually run

Since this PR references an external repo ([kallebysantos/play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests)), please confirm:

- All code snippets are copy-paste runnable
- Import paths (`jsr:`, `npm:`, `https://deno.land/...`) are current and not deprecated
- Deno version requirements are called out if anything requires Deno 1.x vs 2.x

---

### 3. The patch cuts off mid-sentence

The patch ends abruptly with:
> "...a good testing strategy gives you fast "

This is likely a truncation artifact in the PR diff preview, but **please confirm the full sentence is present in the actual file** and the content isn't accidentally cut off in production.

---

### 4. Consider adding a prerequisites/setup section

Readers will benefit from knowing upfront:

- Which version of Deno is required
- Whether `supabase` CLI is needed for integration tests
- Any required environment variables (e.g., `SUPABASE_URL`, `SUPABASE_ANON_KEY`)

A small callout block at the top would help:

```mdx
<Admonition type="note">
These examples require Deno 1.40+ and the Supabase CLI for integration testing.
</Admonition>
```

---

### 5. Mocking strategy — clarify scope

If the docs now cover mocking (e.g., mocking `fetch`, Supabase client, etc.), consider explicitly stating **what is and isn't** being mocked. Developers sometimes mistake mocked integration tests for real integration tests. A brief note on the tradeoffs would add clarity.

---

### 6. Link to the reference repo more prominently

The source repo ([kallebysantos/play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests)) is great supporting material. Consider adding it as a **"See also"** or **"Full example"** callout within the doc body itself, not just in the PR description, so readers can find it.

---

## Minor Nits

- Ensure consistent use of `Edge Functions` (capitalized) vs `edge functions` throughout the new content
- Check that all code blocks have language identifiers (` ```ts `, ` ```bash `, etc.) for syntax highlighting
- If there are `TODO` or placeholder comments in any code snippets, remove them before merging

---

## Summary

This is a really solid docs improvement and the community will benefit from it. The main things to address before merging are: verifying the `#` heading won't duplicate the title, confirming the truncated sentence is complete, and ensuring all code examples are tested and runnable. Once those are confirmed, this looks mergeable! 🎉

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 1]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
 subtitle: 'Writing Unit Tests for Edge Functions using Deno Test'
 ---
 
-Testing is an essential step in the development process to ensure the correctness and performance of your Edge Functions.
+# Testing your Edge Functions
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Since `title` adds the `h1`, I recommend making this an `h2` and so on.

Comment (czenko) [reply apps/docs/content/guides/functions/unit-test.mdx]: Additionally, this duplicates the title. You can name this "Overview" instead.

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 2]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+
+Testing is an essential step in the development process to ensure the correctness, reliability, and performance of your Edge Functions. Because Edge Functions often combine HTTP handling, authentication, database access, and business logic, a good testing strategy gives you fast feedback and high confidence before deploying to production.
+
+In this guide you will learn how to write:
+
+- **Unit tests** for pure business logic (pricing rules, calculations, etc.)
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Parenthetical is typically reserved for defining acronyms. 
```suggestion
- **Unit tests** for pure business logic such as pricing rules, calculations, etc.
```

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 3]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+
+The examples and patterns shown here come directly from the reference implementation in the [play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests) repository and follow the same approaches used internally by Supabase's Edge Functions team.
+
+Deno ships with a fast, native test runner and excellent mocking utilities in `@std/testing`. See the [official Deno testing documentation](https://docs.deno.com/runtime/manual/basics/testing/) for more background.
+
+---
+
+## The example scenario
+
+We will use a realistic Edge Function called `process-ticket` that calculates the final price of a ticket based on the authenticated user's age (loaded from the `profiles` table).
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: - Use present tense
- Use 2nd person singular

For example:
```suggestion
You can use a realistic Edge Function called `process-ticket` that calculates the final price of a ticket based on the authenticated user's age (loaded from the `profiles` table).
```

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 4]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+  return AGE_TIERS.ADULT
+}
+
+export function applyTicketDiscount(price: number, age: number) {
+  const discount = getAgeDiscount(age)
+  return price - price * discount
+}
+```
+
+### Unit tests (`tests/process-ticket/pricing.test.ts`)
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: I'm unsure about this parenthetical here. Headers typically direct the eye but does not give information directly.

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 5]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+Run the integration tests:
+
+```bash
+deno test supabase/functions/tests/process-ticket/index.test.ts --allow-env
+```
 
 ---
 
-## Running Edge Functions locally
+## Why this mocking approach is by mocking `globalThis.fetch`, you:
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Keep headings short. Should probably not include an `:`

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 6]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
 
-To locally test and debug Edge Functions, you can use the Supabase CLI. Let's explore how to run Edge Functions locally using the Supabase CLI:
+- Test the **real** Edge Function code path (no dependency injection needed in production code)
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Again, avoid parenthetical

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 7]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+- Simulate database responses, auth failures, network errors, etc.
+- Keep your production Edge Function clean and focused
+- Still get fast, deterministic tests that don't require a running Supabase instance
 
-1. Ensure that the Supabase server is running by executing the following command:
+This pattern is especially useful when using higher-level helpers like `withSupabase` that create the Supabase client internally.
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Avoid marketing language such as "especially useful". Just say what it does.

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 8]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
-   open .env
-   ```
+```bash
+deno task test
+```
+
+---
 
-4. To run the tests, use the following command in your terminal:
+## Best practices demonstrated in this guide
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: Keep Short. Maybe just "Best practices"

**Review by czenko — COMMENTED** [review]
Here's a style pass. Hope this is helpful.

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 9]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+# Testing your Edge Functions
+
+Testing is an essential step in the development process to ensure the correctness, reliability, and performance of your Edge Functions. Because Edge Functions often combine HTTP handling, authentication, database access, and business logic, a good testing strategy gives you fast feedback and high confidence before deploying to production.
+
+In this guide you will learn how to write:
+
+- **Unit tests** for pure business logic (pricing rules, calculations, etc.)
+- **Integration tests** for the full Edge Function by mocking at the network layer
+
+The examples and patterns shown here come directly from the reference implementation in the [play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests) repository and follow the same approaches used internally by Supabase's Edge Functions team.
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: A bit verbose. Can be simplified.

```suggestion
You can also find the examples in this example repository: [play.supabase.unit-tests](https://github.com/kallebysantos/play.supabase.unit-tests)
```

To "by Supabase's Edge Functions team," it should be implied to be official when this merges in the Supabase org, hopefully.

File: apps/docs/content/guides/functions/unit-test.mdx [hunk 10]
```diff
@@ -5,161 +5,458 @@ description: 'Writing Unit Tests for Edge Functions using Deno Test'
...
+│       ├── utils/
+│       │   └── supabase_env.ts      # Test helpers (env + JWT)
+│       └── process-ticket/
+│           ├── pricing.test.ts      # Unit tests for pricing
+│           └── index.test.ts        # Integration tests with fetch mocking
+├── config.toml
+└── deno.json
+```
+
+> **Note:** In this reference implementation the pricing logic lives inside the function folder (`process-ticket/pricing.ts`). You can also move it to `_shared/` if you want to reuse it across multiple functions.
```
Comment (czenko) [inline apps/docs/content/guides/functions/unit-test.mdx]: I think we have a convention for this.

!!! note

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 8 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: solved at ea9988d5fea4

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 5 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: Solved at 8a22d2276c39

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 4 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: Solved at 1ba749e9a59b

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 3 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: solved at 7cde06c790d0

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 9 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: Solved at 136b85cff0c8

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 1 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: solved at ac6d5f4c5a61

File: apps/docs/content/guides/functions/unit-test.mdx (hunk 2 above)
Comment (kallebysantos) [reply apps/docs/content/guides/functions/unit-test.mdx]: Solved at db7251fc772b

**Review by kallebysantos — APPROVED** [review]

File: examples/edge-functions/supabase/functions/unit-testing/process-ticket/index.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    const { price } = (await req.json()) as Payload
+    if (!price) {
+      return Response.json({ error: 'missing price field' }, { status: 400 })
+    }
+
+    const { data, error: getAgeError } = await ctx.supabase
+      .from('profiles')
+      .select('age')
+      .limit(1)
+      .single()
```
[the remark this answers is not in the document]
Comment (kallebysantos) [reply examples/edge-functions/supabase/functions/unit-testing/process-ticket/index.ts:26]: server sdk `withSupabase` already address RLS

File: examples/edge-functions/supabase/functions/unit-testing/process-ticket/index.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+export type Payload = {
+  price: number
+}
+
+export default {
+  fetch: withSupabase<Database>({ auth: ['user'] }, async (req, ctx) => {
+    const { price } = (await req.json()) as Payload
+    if (!price) {
+      return Response.json({ error: 'missing price field' }, { status: 400 })
+    }
```
[the remark this answers is not in the document]
Comment (kallebysantos) [reply examples/edge-functions/supabase/functions/unit-testing/process-ticket/index.ts:20]: Such validation is not need as its a simple example use-case
